### PR TITLE
Improve HTML+Emd parsing.

### PR DESCRIPTION
**Pull request by bterlson**
This change introduces breaking changes to the parser and emitter
output. The API remains stable.
* Introduces the notion of an HTML tag paragraph which is a paragraph
  that starts with a block HTML tag and ends at either the end of the line
  for normal block-level tags or at the close tag for a small set of
  opaque tags like pre and emu-grammar.
  * Block-level HTML tags are not wrapped in p tags.
  * emd inside block-level tags is allowed but needs to be on a separate
    line. It will be wrapped in p tags (so if you want to use p tags,
    make sure not to break them across multiple lines).
  * Opaque tags cannot contain any emd.
* Adds syntax for headers (One or more `#` characters at the start of a
  line followed by the header text).
* emd.fragment actually parses a fragment with a special case to allow
  paragraph breaks. I still think this capability should be removed.
* Much refactorings, especially to the emitter. The emitter is now
  class-based which will be handy once we expose the AST so clients can
  use custom emitter logic.

### CODE REVIEW

**Comment by bterlson** [comment]
Extensive README updates are probably required, but I'm getting this out in case review suggests any changes :-P

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 language: node_js
 node_js:
-  - iojs-v2.0.0
-  - iojs
+  - "4.0"
+  - "4.1"
```
Comment (ljharb) [inline .travis.yml]: 4.2 has LTS, 5.2 is latest - you may want to use those instead of 4.0 and 4.1.

File: lib/ecmarkdown.js [hunk 2]
```diff
@@ -17,3 +18,9 @@ function parse(emd) {
   const p = new Parser(t);
   return p.parseDocument();
 }
+
+function parseFragment(emd) {
```
Comment (domenic) [inline lib/ecmarkdown.js]: Inline this function?

File: test/paragraph-cases/dunderproto.html [hunk 3]
```diff
@@ -0,0 +1,2 @@
+<p>__proto__ should render verbatim, likewise Object.prototype.__proto__
+</p>
```
Comment (domenic) [inline test/paragraph-cases/dunderproto.html]: I think we should trim trailing whitespace in paragraphs. What do you think?

File: .travis.yml (hunk 1 above)
Comment (domenic) [inline .travis.yml]: just `node_js: latest` will be fine for me; I have no intention of supporting LTS.

**Comment by domenic** [comment]
This is looking great; a bit more minimal than what you showed me yesterday, in all the right ways.

> Adds syntax for headers (One or more # characters at the start of a line followed by the header text).

Add some tests for `#header no whitespace` and `#    header lots of whitespace` and maybe even `#  header trailing whitespace   `? We should trim the whitespace, like CommonMark does, and leave `#header no whitespace` verbatim. I see we already trim `# header single space`.

Also, do you want to validate that we don't go beyond 6 `#`s, and leave verbatim if so? That's what CommonMark does.

When we write up the readme for this we should emphasize that `#` is usually correct, with the TOC outline being created from the `<emu-clause>` structure, but `##` and `###` and friends can be used for special cases of TOC-less headers.

> emd.fragment actually parses a fragment with a special case to allow paragraph breaks. I still think this capability should be removed.

Say more? I would be fine killing `emd.fragment` if you think there are no use cases left for it.

> This change introduces breaking changes to the parser and emitter output.

In addition to the "breaking change" of supporting new things (thus reinterpreting output that would previously be left verbatim), what are the remaining breaking changes that are more on the side of changing previously-supported things? It looks like the big one is wrapping all ols with emu-alg, right? Anything else of note?

File: .travis.yml (hunk 1 above)
Comment (ljharb) [inline .travis.yml]: In that case you'd want `node`, since travis uses `nvm`'s syntax. (unless "latest" is some magical travis thing that proxies to `nvm`). Kind of sad to explicitly drop support for node 4 tho, I'm not sure I see the benefit of dropping it or the cost of keeping it.

Comment (bterlson) [inline .travis.yml]: "4.2" , "5.2" seems fine. Even if we don't want to support LTS, knowing when LTS is broken would be good :)

Comment (ljharb) [inline .travis.yml]: 5.3 just came out a few minutes ago, so s/5.2/5.3 :-p

File: lib/ecmarkdown.js (hunk 2 above)
Comment (bterlson) [inline lib/ecmarkdown.js]: Not sure what you mean? It seems equivalent to parse() and so should be a peer of it, wherever it is.

**Comment by bterlson** [comment]
> Add some tests for #header no whitespace and # header lots of whitespace and maybe even # header trailing whitespace? We should trim the whitespace, like CommonMark does, and leave #header no whitespace verbatim. I see we already trim # header single space.

Ok, am fine requiring whitespace after header.

> Also, do you want to validate that we don't go beyond 6 #s, and leave verbatim if so? That's what CommonMark does.

I guess.

> Say more? I would be fine killing emd.fragment if you think there are no use cases left for it.

The current implementation is better than before, but still not good. You will get odd behavior if you try to parse a fragment that contains block-level elements like headers and such. Probably these should be error conditions? I'm just not sure it's worth maintaining this feature. Without it, consumers would be forced to parse everything as a document and make their own policy decisions for how to handle block-level content (simplest approach is just look inside the first p tag, but depending on scenario they might want something more rigorous).

> In addition to the "breaking change" of supporting new things (thus reinterpreting output that would previously be left verbatim), what are the remaining breaking changes that are more on the side of changing previously-supported things? It looks like the big one is wrapping all ols with emu-alg, right? Anything else of note?

I think the biggest change is how html block tags are parsed. Previously all tags were "inline" tags and so something like `<h1>*format*</h1>` would work fine (although it would be wrapped in p tags). Now *format* is verbatim.

File: test/paragraph-cases/dunderproto.html (hunk 3 above)
Comment (bterlson) [inline test/paragraph-cases/dunderproto.html]: I like keeping it as it makes the output HTML vaguely formatted like the input document :-P Up to you, though, not too hard to remove (although not as simple as a .trim() :-P)

File: .travis.yml (hunk 1 above)
Comment (domenic) [inline .travis.yml]: 4.2 and "node" or whatever latest is (I could swear there was a better syntax for it. Maybe "stable"?)

File: lib/ecmarkdown.js (hunk 2 above)
Comment (domenic) [inline lib/ecmarkdown.js]: Bah you're right I wasn't looking at the whole document.

File: test/paragraph-cases/dunderproto.html (hunk 3 above)
Comment (domenic) [inline test/paragraph-cases/dunderproto.html]: Given what CommonMark does, let's trim it.

**Comment by domenic** [comment]
I'm fine killing fragment.

**Comment by bterlson** [comment]
Pushed updates with a bunch of header tests and made whitespace following the header required. Also moved to an AUTHORS file with both of our names in it :-P

**Comment by bterlson** [comment]
@domenic want to take a stab at the required doc updates? I'm thinking we should remove the grammar and explain-by-example. I'd consider these examples to be the normative definition of this format, with anything not shown in the examples being "implementation defined" for the purposes of other people wanting to write parsers for this format. It will also be much easier to understand for people, probably.

**Comment by domenic** [comment]
Yeah, that sounds good to me. Documentation is the least I can contribute here :)

File: lib/tokenizer.js [hunk 4]
```diff
@@ -1,8 +1,29 @@
...
+  'menu', 'menuitem', 'meta', 'nav', 'noframes', 'ol', 'optgroup', 'option', 'p', 'param', 'section', 'source', 'title',
+  'summary', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'title', 'tr', 'track', 'ul'
+]);
+
+const opaqueTags = new Set([
+  'emu-grammar',
+  'emu-production',
+  'pre',
+  'code',
+  'script'
```
Comment (domenic) [inline lib/tokenizer.js]: style too probably?

File: test/document-cases/html-block-elements.ecmarkdown [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<div>This is a div *no emd*</div>
+*this is emd*
+*so is this*
+
+<section>*no emd here*
+  *but emd is allowed here*
+</section>*but not here*
```
Comment (domenic) [inline test/document-cases/html-block-elements.ecmarkdown]: what about before `</section>` but still on the same line?

**Comment by domenic** [comment]
Here is a new problem I haven't seen before: `` `<stuff>` `` should become `<code>&lt;stuff&gt;</code>`, but becomes `<code><stuff></code>` instead.

**Comment by domenic** [comment]
Here is a test case I would like to see pass, or at least close to it. It came up when I was writing the new readme.

```
<emu-intro>
  # Introduction

  This is a document describing Ecmarkdown.
</emu-intro>

<emu-clause id="c1">
  # Clause 1
  This is paragraph content. You can reference _variables_ and other
  inline formatting in paragraphs too.

  1. Numeric lists are implicitly counted as algorithms.
  2. Everyone loves algorithms.

  This is paragraph content.

  ## Syntax

  Second-level headings are rare in most Ecmarkdown documents, as
  `<emu-clause>` plus first-level headings usually serve better. But
  in some cases you want a heading that does not show up in the table
  of contents, so you can use `##` or `###` syntax.

  <emu-grammar>
    Foo :
      `bar` Baz
  </emu-grammar>

  The contents of `<emu-grammar>`, `<emu-production>`, `<pre>`,
  `<code>` and `<script>` tags are left alone by Ecmarkdown.

  <emu-clause id="sub-clause">
    # Foo(_p1_, _p2_)
    Notice how headers can also use Ecmarkdown inline elements.
  </emu-clause>

  <aside>This content will *not* be processed by Ecmarkdown.

  But this content will be, since it's on a new line.

  It's being on the same line as an opening or closing tag that
  stops processing.</aside>
</emu-clause>
```

should become

```html
<emu-intro>
  <h1>Introduction</h1>
  <p>This is a document describing Ecmarkdown.</p>
</emu-intro>
<emu-clause id="c1">
  <h1>Clause 1</h1>
  <p>This is paragraph content. You can reference <var>variables</var> and other inline formatting in paragraphs too.</p>
  <emu-alg>
    <ol>
      <li>Numeric lists are implicitly counted as algorithms.</li>
      <li>Everyone loves algorithms.</li>
    </ol>
  </emu-alg>
  <p>This is paragraph content.</p>
  <h2>Syntax</h2>
  <p>Second-level headings are rare in most Ecmarkdown documents, as <code>&lt;emu-clause&gt;</code> plus first-level headings usually serve better. But in some cases you want a heading that does not show up in the table of contents, so you can use <code>##</code> or <code>###</code> syntax.</p>
  <emu-grammar>
    Foo :
      `bar` Baz
  </emu-grammar>
  <p>The contents of <code>&lt;emu-grammar&gt;</code>, <code>&lt;emu-production&gt;</code>, <code>&lt;pre&gt;</code>, <code>&lt;code&gt;</code> and <code>&lt;script&gt;</code> tags are left alone by Ecmarkdown.</p>
  <emu-clause id="sub-clause">
    <h1>Foo(<var>p1</var>, <var>p2</var>)</h1>
    <p>Notice how headers can also use Ecmarkdown inline elements.</p>
  </emu-clause>
  <aside>This content will *not* be processed by Ecmarkdown.
    <p>But this content will be, since it's on a new line.</p>
    It's being on the same line as an opening or closing tag that stops processing.</aside>
</emu-clause>
```

File: lib/tokenizer.js (hunk 4 above)
Comment (bterlson) [inline lib/tokenizer.js]: Indeed.

File: test/document-cases/html-block-elements.ecmarkdown (hunk 5 above)
Comment (bterlson) [inline test/document-cases/html-block-elements.ecmarkdown]: In that case it's allowed, but it is not a tag block in that case so the p tag that encloses `*but emd is allowed here*` will also enclose `*but not here*</section>`

**Comment by bterlson** [comment]
Hmm ok, I thought removing Author would make npm pick up AUTHORS file. Maybe it will anyway if contributors is absent?

**Comment by bterlson** [comment]
Alright, things should be much more to your liking after the latest commit. I checked in your readme as a test so you can see its output.

**Comment by domenic** [comment]
Merged as f7da4966041d95657272c65e22245559bfe63c5c. Will send a follow-up README PR.
